octodns/provider/azuredns.py

```diff
@@ -0,0 +1,369 @@
+#
+#
+#
+
+from __future__ import absolute_import, division, print_function, \
+    unicode_literals
+
+from azure.common.credentials import ServicePrincipalCredentials
+from azure.mgmt.dns import DnsManagementClient
+from azure.mgmt.dns.models import ARecord, AaaaRecord, CnameRecord, MxRecord, \
+    SrvRecord, NsRecord, PtrRecord, TxtRecord, Zone
+
+from functools import reduce
+import sys
+import logging
+from ..record import Record
+from .base import BaseProvider
+
+
+class _AzureRecord(object):
+    ''' Wrapper for OctoDNS record.
+        azuredns.py:
+        class: octodns.provider.azuredns._AzureRecord
+        An _AzureRecord is easily accessible to Azure DNS Management library
+        functions and is used to wrap all relevant data to create a record in
+        Azure.
+    '''
+
+    def __init__(self, resource_group, record, delete=False):
+        '''
+            :param resource_group: The name of resource group in Azure
+            :type  resource_group: str
+            :param record: An OctoDNS record
+            :type  record: ..record.Record
+            :param values: Parameters for a record. eg IP address, port, domain
+                           name, etc. Values usually read from record.data
+            :type  values: {'values': [...]} or {'value': [...]}
+
+            :type return: _AzureRecord
+        '''
+        self.resource_group = resource_group
+        self.zone_name = record.zone.name[0:len(record.zone.name) - 1]
+        self.relative_record_set_name = record.name or '@'
+        self.record_type = record._type
+
+        if delete:
+            return
+
+        format_u_s = '' if record._type == 'A' else '_'
+        key_name = '{}{}records'.format(self.record_type, format_u_s).lower()
```

Comment:
If you don't mind can you drop a comment in here about what `key_name` is for. Not knowing Azure I'm not sure if it's something required by its api or if it's an implementation detail of your provider.